synspec.py

```diff
@@ -0,0 +1,50 @@
+from synphot.observation import Observation
+import astropy.units as u
+import numpy as np
+
+
+def from_R(source_spectrum, bandpass, R, waverange, force='none'):
+    """
+    Returns a spectrum with bins defined by a spectral resolution
+    R and a given range of wavelengths. The bin width varies over the
+    range of wavelengths such that the ratio of the wavelength to the
+    bin width at that wavelength is equal to the given R.
+    """
+    first_wl = waverange[0].to(u.angstrom).value
+    binset = np.array([first_wl])
+    binwidth = first_wl / R
+
+    while binset[-1] <= waverange[-1].to(u.angstrom).value:
```

Comment:
First of all, I don't think this can be vectorized easily as `binset[i]` depends on `binset[i - 1]`; and also we don't have to worry about speeding it up until it is proven to be a bottleneck; and even then Cythonizing it might be good enough.

However, I am concerned that this algorithm does not round-trip. For example:
```python
waverange = [5000, 5500] * u.AA
R = 100 
```

Following the math, I think `binset` here is actually bin edges. And I *think* you mean in your docstring that you want the bin's central wavelength divided by its bin width is R, correct? So, having calculated `binset` with this algorithm, my attempt to round trip as follows:

```python
dw = binset[1:] - binset[:-1]  # Bin widths
wcen = binset[:-1] + dw / 2  # Recalculate bin centers
observed_R = wcen / dw
```

And I get 200.5, not 200.

So then, I tried to calculate the bins by hand, knowing the starting wavelength and R (knowing ending wavelength does not help much at the start here, but it lets you know when to stop the bins). However, I got stuck after the first bin. This is because, given that every bin is defined as `wcen / dw = R`, and I know the edge of the first bin, for the second bin, I know neither the `wcen` nor its `dw` in advance, and it seems to me that I have to solve for two unknown at once. Am I going crazy here?

Then I looked around in `specutils` and only found [spectral_resolution()](
https://github.com/astropy/specutils/blob/91041bb0821b8a89e5e15a2fd161ffd46fe5816c/specutils/spectra/spectrum1d.py#L276) that calculates the R given the wavelengths, but not the other way around. This makes me think that maybe requiring `R` to be constant throughout is not practical in the first place?

:thinking: